### ISSUE TITLE
"Warn others: no" setting is ignored on import of a test

### ISSUE THREAD

**Issue by HolgerJeromin** [corona-warn-app/cwa-wishlist#850]
## Avoid duplicates
* [x] Bug is not mentioned in the [FAQ](https://www.coronawarn.app/en/faq/)
* [x] Bug is specific for Android only, for general issues / questions that apply to iOS and Android please raise them in the [documentation repository](https://github.com/corona-warn-app/cwa-documentation)
* [x] Bug is not already reported in another issue

## Technical details
- Device name: Motorola G7 Power
- Android version: 10
- App version: 2.16.2 (still waiting for 2.17)

## Steps to reproduce the issue
1. Scan test QR code (tested with one PCR and two Antigen tests)
2. Set "warn others: no" in import dialog (as this was the test of my child)
3. In the test details (result is not known at this time) I can see "warn others: yes"
4. be happy that you can at least change the setting to "warn others: no" in the test pending dialog

## Expected behaviour
I expected that my first choice is kept.

**Comment by francwalter** [comment]
Hello, I have the problem that I didn't set the warn others to "No" when I got the positive result of PCR of my child, I was a bit in a hurry.

Now the app is warning others from me despite I am negative!
Can I check this later, making the warning others undone?
Is there only the way to reset all data?
Then I have to import all certificates again :(

I miss the feature to import qr codes not for myself but e.g. children. 
I guess till it's implemented it's over anyway with omikron, which is then great anyway ;)
Thank
Frank

**Comment by Ein-Tim** [comment]
@HolgerJeromin There is no option to undo the warning.

**Comment by francwalter** [comment]
> @HolgerJeromin There is no option to undo the warning.

So I guess it is not meant to import PCR-test-QR-codes of others (e.g. children) anyway?

**Comment by Ein-Tim** [comment]
@francwalter 

Not really, no.

**Comment by francwalter** [comment]
> @francwalter 
> 
> Not really, no.

I guess it's no need to put this as feature request now anymore (as Omikron seems to be harmless and one country after another is skipping all COVID-19 restrictions)?
Because I don't see any way to get children's result of PCR test when I just get the QR code. My child's doctor gave me that paper with the QR code and there is information about the app. Doctor told me to use it because he won't be able to tell the result in time, too busy.

2022-09-13: amazing that this app still exists. I uninstalled it months ago, since CoVid-19 is not that dangerous anymore (in combination with vacc. etc.). No masks are worn in the most parts of social life and at the moment only germany still insists on laws about it, lonely in europe. 
I doubt a lot, that this app is still of any use.

**Comment by Ein-Tim** [comment]
@francwalter 

I really understand your problem and I want to point you to https://github.com/corona-warn-app/cwa-wishlist/issues/109 where another user raised the feature request to be able to manage more than one test, however, it was declined.

I opened https://github.com/corona-warn-app/cwa-wishlist/issues/623 which requests to add a checkbox "This is my test" and only if ticked others can be warned.

**Comment by HolgerJeromin** [comment]
> @HolgerJeromin There is no option to undo the warning.

Please read my report (pre test result).

@francwalter kind of highjacked my issue with a completely different issue (post test result).

**Comment by HolgerJeromin** [comment]
Can someone please hide the missleading messages? https://docs.github.com/en/communities/moderating-comments-and-conversations/managing-disruptive-comments#hiding-a-comment

**Comment by dsarkar** [comment]
@HolgerJeromin thanks for the report, we will have a look at this. 
@francwalter Thanks to you for participating here and giving some feedback. However, please, open another issue for your discussion, or if applicable continue in https://github.com/corona-warn-app/cwa-wishlist/issues/623 as suggested above.

**Comment by mtwalli** [comment]
The app now supports family tests 
<img width="569" alt="Screenshot 2022-09-13 at 09 07 11" src="https://user-images.githubusercontent.com/25054729/189833669-baf1dc29-2acf-4bf2-af02-a4539edde9a8.png"> ,but still there is no way to undo the consent once a test is registered. I think this issue should moved to wishlist as it also concerns iOS

**Comment by Ein-Tim** [comment]
> but still there is no way to undo the consent once a test is registered

I think there actually is under both iOS & Android, or not? 

After registering a test I click on the registered test ("Display Test"), scroll to the very bottom, tap on "Consent "Warnt others" granted" & turn the switch off. 

But if I understand @HolgerJeromin's report correctly he selected to not warn others in the Google pop up but CWA still displayed that the Consent to warn others was granted.

Under iOS we have a similar problem which is simply caused by the fact that CWA does not get the consent status back from the API, see https://github.com/corona-warn-app/cwa-app-ios/issues/2840#issuecomment-873559678.

**Comment by mtwalli** [comment]
> > but still there is no way to undo the consent once a test is registered
> 
> 
> 
> I think there actually is under both iOS & Android, or not? 
> 
> 
> 
> After registering a test I click on the registered test ("Display Test"), scroll to the very bottom, tap on "Consent "Warnt others" granted" & turn the switch off. 
> 
> 
> 
> But if I understand @HolgerJeromin's report correctly he selected to not warn others in the Google pop up but CWA still displayed that the Consent to warn others was granted.
> 
> 
> 
> Under iOS we have a similar problem which is simply caused by the fact that CWA does not get the consent status back from the API, see https://github.com/corona-warn-app/cwa-app-ios/issues/2840#issuecomment-873559678.
> 
> 

I did not get where you can change your consent. can you post a screenshot? Before when a positive was registered user had to give OS consent and CWA consent while registering the test once test is registered there is no way where you can revert this choice from test result details screen.

**Comment by Ein-Tim** [comment]
@mtwalli 

> I did not get where you can change your consent. can you post a screenshot?

For sure, I will post a whole screen recording registering a PCR test here, selecting "Don't allow" in the Google pop up to demonstrate the bug @HolgerJeromin is reporting in this issue.

Here's the screen recording:

https://user-images.githubusercontent.com/67682506/189946951-5e62dbc9-b3ca-455c-a8c0-4de22808fe46.mp4

Here's a screenshot from the Google settings.
<img src="https://user-images.githubusercontent.com/67682506/189946970-f7dde95d-7434-4d06-a693-ad1f45b7eec5.png" width="300">

As you see, while CWA always displayed that the consent to warn others was granted, the Google settings always reflected the correct state, not granted (as I selected "Don't allow" in the Google pop up while registering the test).

**Comment by mtwalli** [comment]
@Ein-Tim , I think there is a misunderstanding here. These are `two different` consents and they are not related to each other. 
1- CWA consent 
2- Google ENS consent since `ENS v1.8`  - Rejecting this consent does not mean that CWA consent is disabled. 
Note that Google consent can be shown only once in 5 days see 
[requestPreAuthorizedTemporaryExposureKeyHistory()](https://developers.google.com/android/reference/com/google/android/gms/nearby/exposurenotification/ExposureNotificationClient#requestPreAuthorizedTemporaryExposureKeyHistory()), you would notice that this dialog does not show up if you try to register a new test again after you rejected it for first test, and keys can still be requested from ENS as of pre v1.8 (another consent will show up the one with 14 days )

- if user wish to avoid sharing keys through CWA -> CWA consent should be `disabled` 
- other consent are related to how to get the keys from Google play services (pre 1.8 and after v1.8)

As long as the test is pending , you can turn off/on CWA consent ,but not when it is positive, I thought that even if the test was positive , opt out is needed which is not possible in that case.

See this recording to have a better idea 

https://user-images.githubusercontent.com/25054729/189972004-9a2dfd95-f35a-4e7c-bf06-dc246de00570.mp4

**Comment by Ein-Tim** [comment]
@mtwalli 

Thank you for the clarification, yes indeed, this is really confusing! What exactly does the consent in the CWA itself do?

Help me to understand this using these scenarios:

Scenario 1:
1. User registers test in the app & denies access in Google pop up but CWA consent stays ON
2. Test result comes in positive -> CWA can't automatically warn others in this case but will ask the user to share their keys manually

Scenario 2:
1. User registers test in the app & allows access in the Google pop up but CWA consent is turned OFF
2. Test result comes in positive -> CWA won't (but could) automatically warn others and won't ask the user to share their keys manually

Is this right?

---

Maybe @HolgerJeromin could answer whether what I was describing in https://github.com/corona-warn-app/cwa-wishlist/issues/850#issuecomment-1245603351 is also what he was experiencing?

**Comment by mtwalli** [comment]
@Ein-Tim  that is right. As you already mentioned as long as the test is pending user can change CWA consent. I consider this issue as completed

EDIT: For case 2, I think the App will show that user has a positive test and indication that user can warn others ,later on in the flow user will be asked again for his/her consent

**Comment by Ein-Tim** [comment]
Thanks for all your explanations @mtwalli!

I'll open a wishlist issue requesting to make this all a bit clearer to the average user.

**Comment by HolgerJeromin** [comment]
> Maybe @HolgerJeromin could answer whether what I was describing in https://github.com/corona-warn-app/cwa-wishlist/issues/850#issuecomment-1245603351 is also what he was experiencing?

yes.
Screenshot from your video:

<img width="240" alt="image" src="https://user-images.githubusercontent.com/2410353/190096893-f6e16518-b233-4a4e-81d3-27b57bbd6f22.png">

IMO this is a bug or at least very much feels like a bug (exactly the bug I reported here). You did select "not grant" but CWA shows "granted". "Nobody" knows that there are two layers of consent.
The "CWA consent of the test" should be initialized with the google consent state.

> I consider this issue as completed

I do not. :-)

**Comment by Ein-Tim** [comment]
@HolgerJeromin 

Thanks for confirming you were also confused by this 2 layers of consents thing! I will open a new issue in this repository asking for improving this situation in the course of the next days.

I hope that's OK for you?

**Comment by HolgerJeromin** [comment]
Do what you like. 
But (apart from the derailed discussion starting in comment number 2 which should be [hidden](https://github.com/corona-warn-app/cwa-wishlist/issues/850#issuecomment-1244993189)) I see no point in opening a new one but would reopen this one.

**Comment by Ein-Tim** [comment]
I'll have to leave this decision to one of the repository maintainers.

**Comment by mtwalli** [comment]
@HolgerJeromin  why do you still consider this as a bug?
What you denied is another consent and as @Ein-Tim  shared in his recording you can see this reflected in Google settings (off)
You never denied CWA consent and therefore is still on , you can if you want disable it which is possible. These two consents are different and denying one does not mean that you are denying the other. 

if that is not clear then I agree with @Ein-Tim  a more detailed explanation would be required in this case which is a different topic

**Comment by HolgerJeromin** [comment]
It is clear for me as a programmer after reading this issue tracker.

A random person interacts with CWA and got a question if he wants to share his test result. He denies a popup (which I never remembered as a Non-CWA popup) and reenters the test and sees that his "choice was not remembered".

> You never denied CWA consent

"I was asked something quite similar and answered no, which was not remembered." 

I call it it an [UX](https://en.wikipedia.org/wiki/User_experience) bug.
Could be solved if the "cwa-test default" would be the answer of the user dialog.
